src/Compilers/Core/Portable/Text/CompositeText.cs

```diff
@@ -373,5 +381,181 @@ private static void TrimInaccessibleText(ArrayBuilder<SourceText> segments)
+                    {
+                        var nextSegment = _compositeText.Segments[nextSegmentIndex];
+
+                        // Segments between firstSegmentIndexInclusive and lastSegmentIndexInclusive should have either exactly one line or
+                        // exactly two lines and the second line being empty.
+                        Debug.Assert((nextSegment.Lines.Count == 1) ||
+                                     (nextSegment.Lines.Count == 2 && nextSegment.Lines[1].SpanIncludingLineBreak.IsEmpty));
+
+                        lineLength += nextSegment.Lines[0].SpanIncludingLineBreak.Length;
+                    }
+
+                    if (firstSegmentIndexInclusive != lastSegmentIndexInclusive)
+                    {
+                        var lastSegment = _compositeText.Segments[lastSegmentIndexInclusive];
+
+                        // lastSegment should have at least one line.
+                        Debug.Assert(lastSegment.Lines.Count >= 1);
+
+                        lineLength += lastSegment.Lines[0].SpanIncludingLineBreak.Length;
+                    }
+
+                    var resultLine = TextLine.FromSpanUnsafe(_compositeText, new TextSpan(firstSegmentOffset + firstSegmentTextLine.Start, lineLength));
+
+                    // Assert resultLine only has line breaks in the appropriate locations
+                    Debug.Assert(resultLine.ToString().All(static c => !TextUtilities.IsAnyLineBreakCharacter(c)));
+
+                    return resultLine;
+                }
+            }
+
+            private void GetSegmentIndexRangeContainingLine(int lineNumber, out int firstSegmentIndexInclusive, out int lastSegmentIndexInclusive)
+            {
+                int idx = _segmentLineNumbers.BinarySearch(lineNumber);
+                var binarySearchSegmentIndex = idx >= 0 ? idx : (~idx - 1);
+
+                // Walk backwards starting at binarySearchSegmentIndex to find the earliest segment index that intersects this line number
+                for (firstSegmentIndexInclusive = binarySearchSegmentIndex; firstSegmentIndexInclusive > 0; firstSegmentIndexInclusive--)
+                {
+                    if (_segmentLineNumbers[firstSegmentIndexInclusive] != lineNumber)
+                    {
+                        // This segment doesn't start at the requested line, no need to continue to earlier segments.
+                        break;
+                    }
+
+                    // No need to continue to the previous segment if either:
+                    // 1) it ends in \n or
+                    // 2) if ends in \r and the current segment doesn't start with \n
+                    var previousSegment = _compositeText.Segments[firstSegmentIndexInclusive - 1];
+                    var previousSegmentLastChar = previousSegment[^1];
+                    if (previousSegmentLastChar == '\n')
```

Comment:
so i just realized something.  you're breaking only on \n.  is that correct?  what about other forms of newlines that break up lines?

Reply:
Good catch! Went ahead and made the change to instead use TextUtilities.IsAnyLineBreakCharacter to detect whether a charact is a line break char.